### PR TITLE
⚡️Add `torch.compile` Functionality

### PR DESCRIPTION
**Pull request by Abdol**
This PR integrates PyTorch 2.0's [torch.compile](https://pytorch.org/docs/stable/generated/torch.compile.html) functionality to demonstrate performance improvements in torch code. This PR focuses on adding `torch.compile` to `PatchPredictor`.

**Notes:**
- According to the [documentation](https://pytorch.org/tutorials/intermediate/torch_compile_tutorial.html), noticeable performance can be achieved when using modern NVIDIA GPUs (H100, A100, or V100)

**TODO:**
- [x] Resolve compilation errors related to using `torch.compile` in running models
- [x] Initial config
- [x] Add to patch predictor
- [x] Add to registration
- [x] Add to segmentation
- [x] Test on custom models
- [x] Test on `torch.compile` compatible GPUs

### CODE REVIEW

File: tests/test_wsi_registration.py [hunk 1]
```diff
@@ -21,10 +23,32 @@
 RNG = np.random.default_rng()  # Numpy Random Generator
 
 
-def test_extract_features(dfbr_features: Path) -> None:
+def test_extract_features_time(dfbr_features: Path, test_count: int = 25) -> None:
```
Comment (shaneahmed) [inline tests/test_wsi_registration.py:26]: Is this the right place to add this test?

**Comment by shaneahmed** [comment]
https://github.com/pytorch/pytorch/releases/tag/v2.1.0 PyTorch 2.1 now supports torch.compile in Python 3.11

File: tests/conftest.py [hunk 2]
```diff
@@ -578,3 +580,36 @@ def data_path(tmp_path_factory: pytest.TempPathFactory) -> dict[str, object]:
...
+    """A decorator that times the execution of a function.
+
+    Args:
+        fn (Callable): The function to be timed.
+        args (object): Arguments to be passed to the function.
+
+    Returns:
+        A tuple containing the result of the function
+        and the time taken to execute it in seconds.
+    """
```
Comment (shaneahmed) [inline tests/conftest.py:600]:
```suggestion

    """
```

File: tests/models/test_patch_predictor.py [hunk 3]
```diff
@@ -1226,3 +1227,54 @@ def test_cli_model_multiple_file_mask(remote_sample: Callable, tmp_path: Path) -
...
+    sample_patch2: Path,
+    tmp_path: Path,
+) -> None:
+    """Test torch.compile functionality.
+
+    Args:
+        sample_patch1 (Path): Path to sample patch 1.
+        sample_patch2 (Path): Path to sample patch 2.
+        tmp_path (Path): Path to temporary directory.
+    """
```
Comment (shaneahmed) [inline tests/models/test_patch_predictor.py:1248]:
```suggestion

    """
```

**Comment by shaneahmed** [comment]
Updating PR to fix issues with `codecov` integration.

**Comment by shaneahmed** [comment]
@Abdol With PyTorch 2.3 release, please can you check if it is compatible with Python 3.12?

**Comment by GeorgeBatch** [comment]
@Abdol @shaneahmed, PyTorch 2.4 now supports Python 3.12 for `torch.compile()`: https://github.com/pytorch/pytorch/releases/tag/v2.4.0

**Comment by shaneahmed** [comment]
> I'm getting a DeepSource error related to code outside this PR: [link to error](https://app.deepsource.com/gh/TissueImageAnalytics/tiatoolbox/run/815ca16a-207e-4928-ad40-3647f82a3b1d/python/PYL-E0633?listindex=0). Can anyone have a look and let me know if this needs fixing in this PR (or in another one)? Thanks!

As discussed before you can ignore this error.

**Comment by GeorgeBatch** [comment]
@Abdol, I am happy to try it out in the next few days. What did you mean by setting up the config?

**Comment by GeorgeBatch** [comment]
@Abdol, I only had scripts for feature extraction from patches for now, so that will test patch prediction functionality.

Do you know if I should expect any speedup on NVIDIA A6000 nodes? Or does it only apply to V100, A100, H100 nodes?

If it works well and speeds up the process, I might introduce it into my workflow before the next release. Can you please clarify my question regarding the units and resolution I should use for patch extraction: (issue https://github.com/TissueImageAnalytics/tiatoolbox/issues/874) so I do not waste compute time with incorrectly set parameters?

**Comment by Jiaqi-Lv** [comment]
A100 has just got updated. I will test it on that too.

**Comment by Jiaqi-Lv** [comment]
Currently doing some bench marking using `PatchPredictor` on A100:
Test data is a very large WSI. Each comparison was done 3 times. Models are TIAToolBox pretained models.

Patch size = `224x224`; Number of patches = `29632`
- **_Model_**: `ResNet101-kather100k`; _**Batch size**_: `64`;  **_Compile Method_**: `default`; **_Average Speed Up_**: `1.24±0.11`
- **_Model_**: `wide_resnet101_2-kather100k`; _**Batch size**_: `16`,  **_Compile Method_**: `default`; **_Average Speed Up_**: `1.22±0.03`
- ...

Patch size = `96x96`; Number of patches = `39872`
-  **_Model_**: `resnext101_32x8d-pcam`; _**Batch size**_: `32`;  **_Compile Method_**: `default`; **_Average Speed Up_**: `0.96`
-  **_Model_**: `rwide_resnet101_2-pcam`; _**Batch size**_: `32`;  **_Compile Method_**: `default`; **_Average Speed Up_**: awaiting results
- ...

The first forward pass is always extremely slow with torch.compile, the subsequent forward passes get faster. If the number of iterations are too small torch.compile would not produce speed up.

(Some people say it's best to call `.cuda()` before `.compile()` [https://discuss.pytorch.org/t/torch-compile-before-or-after-cuda/176031/2](url) but there is no official docs about this yet.)

**Comment by Jiaqi-Lv** [comment]
Benchmark for `SemanticSegmentor` using A100
Data: `91424` patches, patch_size=`1024x1024`
- **_Model_**: `fcn_resnet50_unet-bcss`; _**Batch size**_: `8`;  **_Compile Method_**: `default`; **_Average Speed Up_**: `1.3x`; **_Average_time_with_compile_**: 3505s; **_Average_time_without_compile_**: 4558s

File: tiatoolbox/models/architecture/utils.py [hunk 4]
```diff
@@ -2,10 +2,91 @@
...
+        logger.warning(
+            ("torch-compile is currently not supported in Python 3.12+. ",),
+        )
+        return model
+
+    if isinstance(
+        model,
+        torch._dynamo.eval_frame.OptimizedModule,  # skipcq: PYL-W0212 # noqa: SLF001
+    ):
+        logger.warning(
```
Comment (shaneahmed) [inline tiatoolbox/models/architecture/utils.py:83]:
```suggestion
        logger.info(
```

File: tiatoolbox/models/architecture/utils.py [hunk 5]
```diff
@@ -2,10 +2,91 @@
...
+    is_torch_compile_compatible()
+
+    # This check will be removed when torch.compile is supported in Python 3.12+
+    if sys.version_info >= (3, 12):  # pragma: no cover
+        logger.warning(
+            ("torch-compile is currently not supported in Python 3.12+. ",),
+        )
+        return model
+
+    if isinstance(
```
Comment (shaneahmed) [inline tiatoolbox/models/architecture/utils.py:79]:
```suggestion
    if isinstance(  # pragma: no cover
```

File: tiatoolbox/models/architecture/utils.py [hunk 6]
```diff
@@ -2,10 +2,91 @@
...
+        model,
+        torch._dynamo.eval_frame.OptimizedModule,  # skipcq: PYL-W0212 # noqa: SLF001
+    ):
+        logger.warning(
+            ("The model is already compiled. ",),
+        )
+        return model
+
+    return torch.compile(model, mode=mode)
```
Comment (shaneahmed) [inline tiatoolbox/models/architecture/utils.py:88]:
```suggestion
    return torch.compile(model, mode=mode)  # pragma : no cover
```

File: .github/workflows/mypy-type-check.yml [hunk 7]
```diff
@@ -6,7 +6,7 @@ on:
   push:
     branches: [ develop, pre-release, master, main ]
   pull_request:
-    branches: [ develop, pre-release, master, main ]
+    branches: [ develop, pre-release, master, main, enhance-torch-compile ]
```
Comment (shaneahmed) [inline .github/workflows/mypy-type-check.yml:9]:
```suggestion
    branches: [ develop, pre-release, master, main]
```

Please delete this now as this PR is ready.

File: .github/workflows/python-package.yml [hunk 8]
```diff
@@ -8,7 +8,7 @@ on:
     branches: [ develop, pre-release, master, main ]
     tags: v*
   pull_request:
-    branches: [ develop, pre-release, master, main ]
+    branches: [ develop, pre-release, master, main, enhance-torch-compile ]
```
Comment (shaneahmed) [inline .github/workflows/python-package.yml:11]:
```suggestion
    branches: [ develop, pre-release, master, main]
```
Please delete this now as this PR is ready.

File: tiatoolbox/models/architecture/utils.py [hunk 9]
```diff
@@ -2,10 +2,91 @@
...
+    Returns:
+        Callable:
+            Compiled model.
+
+    """
+    if mode == "disable":
+        return model
+
+    # Check if GPU is compatible with torch.compile
+    is_torch_compile_compatible()
```
Comment (shaneahmed) [inline tiatoolbox/models/architecture/utils.py:70]: shouldn't this raise an error?

File: tiatoolbox/models/engine/semantic_segmentor.py [hunk 10]
```diff
@@ -619,12 +623,13 @@ def get_coordinates(
...
         """
-        (patch_inputs, patch_outputs) = PatchExtractor.get_coordinates(
+        results = PatchExtractor.get_coordinates(
+            patch_output_shape=ioconfig.patch_output_shape,
             image_shape=image_shape,
             patch_input_shape=ioconfig.patch_input_shape,
-            patch_output_shape=ioconfig.patch_output_shape,
             stride_shape=ioconfig.stride_shape,
         )
+        patch_inputs, patch_outputs = results[0], results[1]
```
Comment (shaneahmed) [inline tiatoolbox/models/engine/semantic_segmentor.py:632]: This is redundant. On the next line you can do 

`return results[0], results[1]`

File: tiatoolbox/models/engine/semantic_segmentor.py [hunk 11]
```diff
@@ -619,13 +623,13 @@ def get_coordinates(
             >>> segmentor.get_coordinates = func
 
         """
-        (patch_inputs, patch_outputs) = PatchExtractor.get_coordinates(
+        results = PatchExtractor.get_coordinates(
+            patch_output_shape=ioconfig.patch_output_shape,
             image_shape=image_shape,
             patch_input_shape=ioconfig.patch_input_shape,
-            patch_output_shape=ioconfig.patch_output_shape,
```
Comment (shaneahmed) [inline tiatoolbox/models/engine/semantic_segmentor.py:625]: Why this has been deleted?

File: tiatoolbox/models/architecture/utils.py (hunk 9 above)
Comment (shaneahmed) [reply tiatoolbox/models/architecture/utils.py:70]: In that case, `is_torch_compile_compatible()` should have no return as this just raises a warning.

File: tiatoolbox/models/architecture/utils.py [hunk 12]
```diff
@@ -2,10 +2,91 @@
...
 import torch
 from torch import nn
 
+from tiatoolbox import logger
+
+
+def is_torch_compile_compatible() -> bool:
+    """Check if the current GPU is compatible with torch-compile.
+
+    Returns:
```
Comment (shaneahmed) [inline tiatoolbox/models/architecture/utils.py:18]:
```suggestion
    Raises:
        Warning etc if ... [please fill]
```

File: tiatoolbox/models/architecture/utils.py [hunk 13]
```diff
@@ -2,10 +2,91 @@
...
+from typing import Callable
+
 import numpy as np
 import torch
 from torch import nn
 
+from tiatoolbox import logger
+
+
+def is_torch_compile_compatible() -> bool:
```
Comment (shaneahmed) [inline tiatoolbox/models/architecture/utils.py:15]:
```suggestion
def is_torch_compile_compatible() -> NoReturn:
```

File: tiatoolbox/models/architecture/utils.py [hunk 14]
```diff
@@ -2,10 +2,91 @@
...
+    if torch.cuda.is_available():  # pragma: no cover
+        device_cap = torch.cuda.get_device_capability()
+        if device_cap not in ((7, 0), (8, 0), (9, 0)):
+            logger.warning(
+                "GPU is not compatible with torch.compile. "
+                "Compatible GPUs include NVIDIA V100, A100, and H100. "
+                "Speedup numbers may be lower than expected.",
+                stacklevel=2,
+            )
+            return False
```
Comment (shaneahmed) [inline tiatoolbox/models/architecture/utils.py:33]:
```suggestion
```

File: tiatoolbox/models/architecture/utils.py [hunk 15]
```diff
@@ -2,10 +2,91 @@
...
+    else:
+        logger.warning(
+            "No GPU detected or cuda not installed, "
+            "torch.compile is only supported on selected NVIDIA GPUs. "
+            "Speedup numbers may be lower than expected.",
+            stacklevel=2,
+        )
+        return False
```
Comment (shaneahmed) [inline tiatoolbox/models/architecture/utils.py:41]:
```suggestion
```

File: tiatoolbox/models/architecture/utils.py [hunk 16]
```diff
@@ -2,10 +2,91 @@
...
+
+    return True  # pragma: no cover
```
Comment (shaneahmed) [inline tiatoolbox/models/architecture/utils.py:43]:
```suggestion
```

File: .github/workflows/python-package.yml [hunk 17]
```diff
@@ -8,7 +8,7 @@ on:
     branches: [ develop, pre-release, master, main ]
     tags: v*
   pull_request:
-    branches: [ develop, pre-release, master, main ]
+    branches: [ develop, pre-release, master, main]
```
Comment (shaneahmed) [inline .github/workflows/python-package.yml:11]:
```suggestion
    branches: [ develop, pre-release, master, main ]
```

File: .github/workflows/mypy-type-check.yml [hunk 18]
```diff
@@ -6,7 +6,7 @@ on:
   push:
     branches: [ develop, pre-release, master, main ]
   pull_request:
-    branches: [ develop, pre-release, master, main ]
+    branches: [ develop, pre-release, master, main]
```
Comment (shaneahmed) [inline .github/workflows/mypy-type-check.yml:9]:
```suggestion
    branches: [ develop, pre-release, master, main ]
```

**Review by shaneahmed — COMMENTED** [review]
Thanks @Abdol  and @Jiaqi-Lv Just a couple of comments about `is_torch_compile_compatible`. The rest looks fine to me. If you agree with my comments then please make the changes. I will approve and merge it then.

File: tiatoolbox/models/engine/semantic_segmentor.py (hunk 11 above)
Comment (Jiaqi-Lv) [reply tiatoolbox/models/engine/semantic_segmentor.py:625]: It was moved to the top

File: tiatoolbox/models/architecture/utils.py [hunk 19]
```diff
@@ -2,10 +2,91 @@
...
+    *,
+    mode: str = "default",
+) -> Callable:
+    """A decorator to compile a model using torch-compile.
+
+    Args:
+        model (torch.nn.Module):
+            Model to be compiled.
+        mode (str):
+            Mode to be used for torch-compile. Available modes are
```
Comment (shaneahmed) [inline tiatoolbox/models/architecture/utils.py:57]: @Jiaqi-Lv Can we add a description what these modes are?

**Pull request merged**